### ISSUE TITLE
The 'impossible' happened. Compile time panic which I think is related to associated types.

### ISSUE THREAD

**Issue by JustusAdam** [typelead/eta#521]
I got this error 

```
[3 of 5] Compiling Ohua.Compat.JVM.Marshal ( src/eta/Ohua/Compat/JVM/Marshal.hs, dist/build/Ohua/Compat/JVM/Marshal.jar ) [Ohua.Compile changed]
[4 of 5] Compiling Ohua.Compat.JVM.ToALang ( src/eta/Ohua/Compat/JVM/ToALang.hs, dist/build/Ohua/Compat/JVM/ToALang.jar ) [Ohua.Compat.JVM.Marshal changed]
[5 of 5] Compiling Ohua.Compat.JVM.Compiler ( src/eta/Ohua/Compat/JVM/Compiler.hs, dist/build/Ohua/Compat/JVM/Compiler.jar )

<no location info>:
    eta: panic! (the 'impossible' happened)
  (Eta version 0.0.9b1):
        resultWrapper NativeType QualifiedBinding

Please report this as a Eta bug: http://github.com/typelead/eta/issues
```

When compiling the [ohua-jvm-integration](https://github.com/ohua-dev/ohua-jvm-integration) project.

## Notes

I use a sandbox for the project and the packages are frozen with the `cabal.config` file.
The `ohua-core` dependency is added via `etlas sandbox add-source`.

**Comment by rahulmutt** [comment]
This looks to be an issue with declaring the result of a foreign import, I'll reproduce it locally and let you know what the issue is.

**Comment by rahulmutt** [comment]
I got this when I compiled `master` of `ohua-jvm-integration`:
```
src/eta/Ohua/Compat/JVM/Compiler.hs:51:18:
    Ambiguous occurrence ‘spliceEnv’
    It could refer to either ‘Ohua.Compat.JVM.Compiler.spliceEnv’,
                             defined at src/eta/Ohua/Compat/JVM/Compiler.hs:38:1
                          or ‘Ohua.DFGraph.spliceEnv’,
                             imported from ‘Ohua.DFGraph’ at src/eta/Ohua/Compat/JVM/Compiler.hs:14:1-19
```

Can you share what changes you made to fix it?

**Comment by JustusAdam** [comment]
Sorry, I had forgotten to push a change I made. If you pull `ohua-jvm-integration` again it should work now.

**Comment by sertel** [comment]
@rahulmutt This is a blocker for our project. Can you please have a look at this again?
Thanks a lot!

**Comment by rahulmutt** [comment]
Apologise for the late response! The problem is the use of associated type families in a FFI import - this is not supported yet. The solution for now to unblock your project is to inline the the use of `NativeType` in your FFI imports. In the particular case,

```haskell
foreign import java unsafe "@interface resolve" linkerResolveUnqualified ::
  String -> Java IsLinker (Maybe JString)
```

Will fix it. I'll file a new issue about supporting type families.

**Comment by rahulmutt** [comment]
I revisited this an it turns out that the problem is not type families exactly, but the use of type families in a nested manner like `Java IsLinker (Maybe (NativeType QualifiedBinding))`. I added an extra step to the FFI typechecking to see through nested type families as well so you are free to go back to using the `NativeType` type family in your codebase. I'll be doing a release in a couple days and will update this issue once it's ready.
